### PR TITLE
Fixes #22338: Refactor some code and changes API url for compliance by directive

### PR DESCRIPTION
**Pull request by ElaadF**
https://issues.rudder.io/issues/22338

### CODE REVIEW

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala [hunk 1]
```diff
@@ -259,6 +259,7 @@ trait StartsAtVersion13 extends EndpointSchema { val versions = ApiV.From(13) }
 trait StartsAtVersion14 extends EndpointSchema { val versions = ApiV.From(14) }
 trait StartsAtVersion15 extends EndpointSchema { val versions = ApiV.From(15) }
 trait StartsAtVersion16 extends EndpointSchema { val versions = ApiV.From(16) }
+trait StartsAtVersion17 extends EndpointSchema { val versions = ApiV.From(17) }
```
Comment (amousset) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala:262]: this also need to be changed in the API doc

Comment (ElaadF) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala:262]: there is no API doc for the moment, problem solved :)

Comment (ElaadF) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala:262]: wrong version issue reported by @ncharles [here](https://github.com/Normation/rudder/pull/4626#discussion_r1112837815)

**Comment by ElaadF** [comment]
Commit modified

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala (hunk 1 above)
Comment (amousset) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/ApiDatastructures.scala:262]: yes but you need to change version to 17 in the doc too

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Comment by VinceMacBuche** [comment]
Commit modified

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala [hunk 2]
```diff
@@ -160,24 +160,24 @@ object ComplianceApi       extends ApiModuleProvider[ComplianceApi] {
...
     val (action, path) = GET / "compliance" / "directives"
     val dataContainer  = Some("directivesCompliance")
   }
 
-  final case object ExportDirectiveComplianceCSV extends ComplianceApi with OneParam with StartsAtVersion10 with SortIndex {
+  final case object ExportDirectiveComplianceCSV extends ComplianceApi with OneParam with StartsAtVersion17 with SortIndex {
     val z              = implicitly[Line].value
     val description    = "Get a directive's compliance to CSV format"
-    val (action, path) = GET / "compliance" / "directives" / "export" / "{id}"
+    val (action, path) = GET / "compliance" / "directives" / "{id}"
```
[the remark this answers is not in the document]
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala]: Yes i meant that csv and json should be the same endpoint, with a default format value to be json. I made a second commit with refactoring

**Review by VinceMacBuche — APPROVED** [review]

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 3]
```diff
@@ -246,35 +207,32 @@ class ComplianceApi(
...
+        format match {
+          case ComplianceFormat.CSV  =>
+            PlainTextResponse(directive.toCsv.mkString("\n"))
+          case ComplianceFormat.JSON =>
+            val json = directive.toJson(
+              level.getOrElse(10),
+              precision.getOrElse(CompliancePrecision.Level2)
+            ) // by default, all details are displayed
+            val t4 = System.currentTimeMillis
+            TimingDebugLogger.trace(s"API DirectiveCompliance - serialize to json in ${t4 - t3} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:231]: I have 5,5s on this part to show compliance per directive on load platform

**Pull request merged**